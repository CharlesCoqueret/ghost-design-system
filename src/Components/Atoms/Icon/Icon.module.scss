.icon {
  // Disable font smoothing for performance reason
  -webkit-font-smoothing: unset;
}
