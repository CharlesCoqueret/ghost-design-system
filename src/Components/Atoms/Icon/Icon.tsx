import React, { ReactElement } from 'react';
import { FontAwesomeIcon, FontAwesomeIconProps } from '@fortawesome/react-fontawesome';
import classnames from 'classnames';

import Loader from './icons/Loader';

const Icon = ({ className, ...props }: FontAwesomeIconProps): ReactElement => {
  if (props.icon && Array.isArray(props.icon) && props.icon.length === 2) {
    if (props.icon[1] === 'spinner') return <Loader {...props} size={props.size} className={className} />;
  }
  return <FontAwesomeIcon {...props} className={classnames(className, 'icon')} />;
};

export default Icon;
