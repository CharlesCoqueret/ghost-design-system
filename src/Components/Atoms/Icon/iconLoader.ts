import { dom, library } from '@fortawesome/fontawesome-svg-core';

import {
  faArrowDownToLine,
  faArrowLeft,
  faArrowsToLine,
  faArrowUpFromLine,
  faArrowsLeftRightToLine,
  faArrowCircleRight, // deprecated
  faBallotCheck,
  faBell,
  faBriefcase,
  faCalendarDay,
  faChartColumn,
  faCheck,
  faChevronDown,
  faChevronLeft,
  faChevronRight,
  faColumns3,
  faEllipsis,
  faExclamationTriangle,
  faExternalLink,
  faEye,
  faFileAlt,
  faFilter,
  faGear,
  faHighlighter,
  faInbox,
  faInfoCircle,
  faListOl,
  faListUl,
  faLoader,
  faLongArrowAltDown,
  faLongArrowAltLeft,
  faLongArrowAltUp,
  faMagnifyingGlass,
  faMessageExclamation,
  faMinus,
  faPaperclip,
  faPaperPlane,
  faPen,
  faPipe,
  faPlus,
  faQuestionCircle,
  faSave,
  faScaleBalanced,
  faSort as faSortLight, // deprecated
  faSquare as faSquareLight,
  faTable,
  faTextSize,
  faTimes,
  faTimesCircle,
  faTrashAlt,
  faUndo,
  faUserCircle,
  faUserLock,
  faUserUnlock,
  faWindowMaximize,
} from '@fortawesome/pro-light-svg-icons';

import {
  // Text editor
  faBold,
  faHorizontalRule,
  faImage,
  faIndent,
  faItalic,
  faLink,
  faOutdent,
  faPalette,
  faStrikethrough,
  faUnderline,
} from '@fortawesome/pro-regular-svg-icons';

// Button and checkbox
import {
  faCaretDown,
  faCircleCheck,
  faCircleXmark,
  faSquare,
  faSquareCheck,
  faSort,
  faSortDown,
  faSortUp,
} from '@fortawesome/pro-solid-svg-icons';

const loadIcons = (): void => {
  library.add(
    faArrowCircleRight,
    faArrowDownToLine,
    faArrowLeft,
    faArrowsLeftRightToLine,
    faArrowsToLine,
    faArrowUpFromLine,
    faBallotCheck,
    faBell,
    faBriefcase,
    faCalendarDay,
    faChartColumn,
    faCheck,
    faChevronDown,
    faChevronLeft,
    faChevronRight,
    faColumns3,
    faMessageExclamation,
    faEllipsis,
    faExclamationTriangle,
    faExternalLink,
    faEye,
    faFileAlt,
    faFilter,
    faGear,
    faInbox,
    faInfoCircle,
    faLoader,
    faLongArrowAltDown,
    faLongArrowAltLeft,
    faLongArrowAltUp,
    faMagnifyingGlass,
    faMinus,
    faPaperclip,
    faPaperPlane,
    faPen,
    faPipe,
    faPlus,
    faQuestionCircle,
    faSave,
    faScaleBalanced,
    faSort,
    faSortLight,
    faSortDown,
    faSortUp,
    faSquareLight,
    faTimes,
    faTimesCircle,
    faTrashAlt,
    faUndo,
    faUserCircle,
    faUserLock,
    faUserUnlock,
    faWindowMaximize,

    // text editor
    faBold,
    faHighlighter,
    faHorizontalRule,
    faImage,
    faIndent,
    faItalic,
    faLink,
    faListOl,
    faListUl,
    faOutdent,
    faPalette,
    faStrikethrough,
    faTable,
    faTextSize,
    faUnderline,

    // button
    faCaretDown,
    faSquare,
    faSquareCheck,

    // toast
    faCircleXmark,
    faCircleCheck,
  );

  dom.watch();
};

export default loadIcons;
