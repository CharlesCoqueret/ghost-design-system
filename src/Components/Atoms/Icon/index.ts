export { default as loadIcons } from './iconLoader';
export { default as Icon } from './Icon';
