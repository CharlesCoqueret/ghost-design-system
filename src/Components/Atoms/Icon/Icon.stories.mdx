<!-- Iconography.stories.mdx -->

import { Meta, Title, IconGallery, IconItem } from '@storybook/addon-docs/';

import Icon from './Icon';

<Meta title='Atom/Icon' />

# Icons with fal prefix

- ['fal', icon name]

<IconGallery>
  <IconItem name='arrow-down-to-line'>
    <Icon icon={['fal', 'arrow-down-to-line']} />
  </IconItem>
  <IconItem name='arrow-left'>
    <Icon icon={['fal', 'arrow-left']} />
  </IconItem>
  <IconItem name='arrow-up-from-line'>
    <Icon icon={['fal', 'arrow-up-from-line']} />
  </IconItem>
  <IconItem name='arrows-left-right-to-line'>
    <Icon icon={['fal', 'arrows-left-right-to-line']} />
  </IconItem>
  <IconItem name='arrows-to-line'>
    <Icon icon={['fal', 'arrows-to-line']} />
  </IconItem>
  <IconItem name='ballot-check'>
    <Icon icon={['fal', 'ballot-check']} />
  </IconItem>
  <IconItem name='bell'>
    <Icon icon={['fal', 'bell']} />
  </IconItem>
  <IconItem name='briefcase'>
    <Icon icon={['fal', 'briefcase']} />
  </IconItem>
  <IconItem name='calendar-day'>
    <Icon icon={['fal', 'calendar-day']} />
  </IconItem>
  <IconItem name='chart-column'>
    <Icon icon={['fal', 'chart-column']} />
  </IconItem>
  <IconItem name='check'>
    <Icon icon={['fal', 'check']} />
  </IconItem>
  <IconItem name='chevron-down'>
    <Icon icon={['fal', 'chevron-down']} />
  </IconItem>
  <IconItem name='chevron-left'>
    <Icon icon={['fal', 'chevron-left']} />
  </IconItem>
  <IconItem name='chevron-right'>
    <Icon icon={['fal', 'chevron-right']} />
  </IconItem>
  <IconItem name='columns-3'>
    <Icon icon={['fal', 'columns-3']} />
  </IconItem>
  <IconItem name='ellipsis'>
    <Icon icon={['fal', 'ellipsis']} />
  </IconItem>
  <IconItem name='exclamation-triangle'>
    <Icon icon={['fal', 'exclamation-triangle']} />
  </IconItem>
  <IconItem name='external-link'>
    <Icon icon={['fal', 'external-link']} />
  </IconItem>
  <IconItem name='eye'>
    <Icon icon={['fal', 'eye']} />
  </IconItem>
  <IconItem name='file-alt'>
    <Icon icon={['fal', 'file-alt']} />
  </IconItem>
  <IconItem name='filter'>
    <Icon icon={['fal', 'filter']} />
  </IconItem>
  <IconItem name='gear'>
    <Icon icon={['fal', 'gear']} />
  </IconItem>
  <IconItem name='highlighter'>
    <Icon icon={['fal', 'highlighter']} />
  </IconItem>
  <IconItem name='inbox'>
    <Icon icon={['fal', 'inbox']} />
  </IconItem>
  <IconItem name='info-circle'>
    <Icon icon={['fal', 'info-circle']} />
  </IconItem>
  <IconItem name='list-ol'>
    <Icon icon={['fal', 'list-ol']} />
  </IconItem>
  <IconItem name='list-ul'>
    <Icon icon={['fal', 'list-ul']} />
  </IconItem>
  <IconItem name='loader'>
    <Icon icon={['fal', 'loader']} />
  </IconItem>
  <IconItem name='long-arrow-alt-down'>
    <Icon icon={['fal', 'long-arrow-alt-down']} />
  </IconItem>
  <IconItem name='long-arrow-alt-left'>
    <Icon icon={['fal', 'long-arrow-alt-left']} />
  </IconItem>
  <IconItem name='long-arrow-alt-up'>
    <Icon icon={['fal', 'long-arrow-alt-up']} />
  </IconItem>
  <IconItem name='magnifying-glass'>
    <Icon icon={['fal', 'magnifying-glass']} />
  </IconItem>
  <IconItem name='message-exclamation'>
    <Icon icon={['fal', 'message-exclamation']} />
  </IconItem>
  <IconItem name='minus'>
    <Icon icon={['fal', 'minus']} />
  </IconItem>
  <IconItem name='paperclip'>
    <Icon icon={['fal', 'paperclip']} />
  </IconItem>
  <IconItem name='paper-plane'>
    <Icon icon={['fal', 'paper-plane']} />
  </IconItem>
  <IconItem name='pen'>
    <Icon icon={['fal', 'pen']} />
  </IconItem>
  <IconItem name='pipe'>
    <Icon icon={['fal', 'pipe']} />
  </IconItem>
  <IconItem name='plus'>
    <Icon icon={['fal', 'plus']} />
  </IconItem>
  <IconItem name='question-circle'>
    <Icon icon={['fal', 'question-circle']} />
  </IconItem>
  <IconItem name='save'>
    <Icon icon={['fal', 'save']} />
  </IconItem>
  <IconItem name='scale-balanced'>
    <Icon icon={['fal', 'scale-balanced']} />
  </IconItem>
  <IconItem name='spinner'>
    <Icon icon={['fal', 'spinner']} />
  </IconItem>
  <IconItem name='square'>
    <Icon icon={['fal', 'square']} />
  </IconItem>
  <IconItem name='table'>
    <Icon icon={['fal', 'table']} />
  </IconItem>
  <IconItem name='text-size'>
    <Icon icon={['fal', 'text-size']} />
  </IconItem>
  <IconItem name='times'>
    <Icon icon={['fal', 'times']} />
  </IconItem>
  <IconItem name='times-circle'>
    <Icon icon={['fal', 'times-circle']} />
  </IconItem>
  <IconItem name='trash-alt'>
    <Icon icon={['fal', 'trash-alt']} />
  </IconItem>
  <IconItem name='undo'>
    <Icon icon={['fal', 'undo']} />
  </IconItem>
  <IconItem name='user-circle'>
    <Icon icon={['fal', 'user-circle']} />
  </IconItem>
  <IconItem name='user-lock'>
    <Icon icon={['fal', 'user-lock']} />
  </IconItem>
  <IconItem name='user-unlock'>
    <Icon icon={['fal', 'user-unlock']} />
  </IconItem>
  <IconItem name='window-maximize'>
    <Icon icon={['fal', 'window-maximize']} />
  </IconItem>
</IconGallery>

# Icons with far prefix

- ['far', icon name]

<IconGallery>
  <IconItem name='bold'>
    <Icon icon={['far', 'bold']} />
  </IconItem>
  <IconItem name='horizontal-rule'>
    <Icon icon={['far', 'horizontal-rule']} />
  </IconItem>
  <IconItem name='image'>
    <Icon icon={['far', 'image']} />
  </IconItem>
  <IconItem name='indent'>
    <Icon icon={['far', 'indent']} />
  </IconItem>
  <IconItem name='italic'>
    <Icon icon={['far', 'italic']} />
  </IconItem>
  <IconItem name='link'>
    <Icon icon={['far', 'link']} />
  </IconItem>
  <IconItem name='outdent'>
    <Icon icon={['far', 'outdent']} />
  </IconItem>
  <IconItem name='palette'>
    <Icon icon={['far', 'palette']} />
  </IconItem>
  <IconItem name='strikethrough'>
    <Icon icon={['far', 'strikethrough']} />
  </IconItem>
  <IconItem name='underline'>
    <Icon icon={['far', 'underline']} />
  </IconItem>
</IconGallery>

# Icons with fas prefix

- ['fas', icon name]

<IconGallery>
  <IconItem name='caret-down'>
    <Icon icon={['fas', 'caret-down']} />
  </IconItem>
  <IconItem name='circle-check'>
    <Icon icon={['fas', 'circle-check']} />
  </IconItem>
  <IconItem name='circle-xmark'>
    <Icon icon={['fas', 'circle-xmark']} />
  </IconItem>
  <IconItem name='sort'>
    <Icon icon={['fas', 'sort']} />
  </IconItem>
  <IconItem name='sort-down'>
    <Icon icon={['fas', 'sort-down']} />
  </IconItem>
  <IconItem name='sort-up'>
    <Icon icon={['fas', 'sort-up']} />
  </IconItem>
  <IconItem name='square'>
    <Icon icon={['fas', 'square']} />
  </IconItem>
  <IconItem name='square-check'>
    <Icon icon={['fas', 'square-check']} />
  </IconItem>
</IconGallery>
