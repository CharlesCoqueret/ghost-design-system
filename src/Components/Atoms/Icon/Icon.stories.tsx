import React from 'react';
import { IconProp } from '@fortawesome/fontawesome-svg-core';
import Icon from './Icon';

export default {
  title: 'Atom/Icon',
};

const Template = (args: { icons: Array<IconProp> }) => {
  const { icons } = args;
  return (
    <div
      style={{
        display: 'grid',
        gridTemplateColumns: 'repeat(3, 1fr)',
        gridGap: '20px',
        gridAutoRows: 'minmax(100px, auto)',
      }}>
      {icons.map((icon, index) => {
        const spin = index % 7 == 0;
        return (
          <div key={icon.toString()} style={{ display: 'flex', flexDirection: 'column', margin: 'auto' }}>
            <Icon icon={icon} size='2x' spin={spin} />
            <div style={{ fontWeight: '400', fontSize: '14px', margin: '10px auto' }}>{`${icon[0]} ${icon[1]}`}</div>
            <div style={{ fontWeight: '300', fontSize: '10px', margin: 'auto' }}>{spin && '(Spinning)'}</div>
          </div>
        );
      })}
    </div>
  );
};

export const List = Template.bind({});
List.args = {
  icons: [
    ['fal', 'arrow-to-bottom'],
    ['fal', 'arrow-left'],
    ['fal', 'balance-scale'],
    ['fal', 'ballot-check'],
    ['fal', 'bell'],
    ['fal', 'briefcase'],
    ['fal', 'chart-bar'],
    ['fal', 'check'],
    ['fal', 'chevron-down'],
    ['fal', 'chevron-left'],
    ['fal', 'chevron-right'],
    ['fal', 'cog'],
    ['fal', 'comment-alt-exclamation'],
    ['fal', 'edit'],
    ['fal', 'ellipsis-h'],
    ['fal', 'exclamation-circle'],
    ['fal', 'exclamation-triangle'],
    ['fal', 'external-link'],
    ['fal', 'eye'],
    ['fal', 'file-alt'],
    ['fal', 'file-medical-alt'],
    ['fal', 'file-powerpoint'],
    ['fal', 'filter'],
    ['fal', 'hand-paper'],
    ['fal', 'highlighter'],
    ['fal', 'inbox'],
    ['fal', 'info-circle'],
    ['fal', 'list-ol'],
    ['fal', 'list-ul'],
    ['fal', 'long-arrow-alt-down'],
    ['fal', 'long-arrow-alt-left'],
    ['fal', 'long-arrow-alt-up'],
    ['fal', 'magnifying-glass'],
    ['fal', 'minus'],
    ['fal', 'paperclip'],
    ['fal', 'paper-plane'],
    ['fal', 'pen'],
    ['fal', 'pipe'],
    ['fal', 'plus'],
    ['fal', 'question-circle'],
    ['fal', 'save'],
    ['fal', 'spinner'],
    ['fal', 'square'],
    ['fal', 'table'],
    ['fal', 'text-size'],
    ['fal', 'times'],
    ['fal', 'times-circle'],
    ['fal', 'trash-alt'],
    ['fal', 'undo'],
    ['fal', 'upload'],
    ['fal', 'user-circle'],
    ['fal', 'user-lock'],
    ['fal', 'user-unlock'],
    ['far', 'bold'],
    ['far', 'horizontal-rule'],
    ['far', 'image'],
    ['far', 'indent'],
    ['far', 'italic'],
    ['far', 'link'],
    ['far', 'outdent'],
    ['far', 'palette'],
    ['far', 'strikethrough'],
    ['far', 'underline'],
    ['fas', 'caret-down'],
    ['fas', 'sort'],
    ['fas', 'sort-down'],
    ['fas', 'sort-up'],
    ['fas', 'square'],
    ['fas', 'square-check'],
  ],
};
