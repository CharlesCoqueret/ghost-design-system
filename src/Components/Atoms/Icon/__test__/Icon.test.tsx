import React from 'react';
import { render } from '@testing-library/react';

import Icon from '../Icon';

describe('Icon component', () => {
  it('Icon renders', async () => {
    const container = render(<Icon icon={['fal', 'cog']} className={'CLASSNAME'} />);
    expect(container).toMatchSnapshot();
  });
});
