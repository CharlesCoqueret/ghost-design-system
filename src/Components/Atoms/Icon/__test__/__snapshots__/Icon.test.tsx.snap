// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component Icon renders 1`] = `
<div>
  <i
    class="CLASSNAME icon"
    icon="fal,cog"
  />
</div>
`;
